File: backend/routers/suppliers.py
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File, Form
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, desc
from sqlalchemy.orm import joinedload
from typing import List, Optional
import os
import uuid
from datetime import datetime

from database import get_db, Supplier, SupplyReceipt, SupplierPayment, Employee
from core import get_current_user, verify_password, verify_approver
from pydantic import BaseModel

from routers.audit import log_action
from utils.backup import UPLOAD_DIR

router = APIRouter(prefix="/suppliers", tags=["suppliers"])

# Nakladnoy fayli uchun ruxsat etilgan kengaytmalar va hajm chegarasi.
ALLOWED_INVOICE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp", ".gif", ".pdf"}
MAX_INVOICE_BYTES = 10 * 1024 * 1024  # 10 MB

# --- Schemas ---
class SupplierBase(BaseModel):
    name: str
    phone: Optional[str] = None
    address: Optional[str] = None

class SupplierCreate(SupplierBase):
    pass

class SupplierOut(SupplierBase):
    id: int
    balance: float
    created_at: datetime

    class Config:
        from_attributes = True

class ReceiptOut(BaseModel):
    id: int
    supplier_id: int
    total_amount: float
    invoice_image: Optional[str] = None
    date: datetime
    note: Optional[str] = None

    class Config:
        from_attributes = True

class PaymentOut(BaseModel):
    id: int
    supplier_id: int
    amount: float
    payment_method: str
    date: datetime
    note: Optional[str] = None

    class Config:
        from_attributes = True

# --- Endpoints ---

async def verify_confirming_employee(db: AsyncSession, requester: Employee,
                                     username: str, password: str) -> Employee:
    """Firma bilan pul harakatini TASDIQLOVCHI xodim.

    Ilgari bu funksiya na rolni, na tasdiqlovchi BOSHQA odam ekanini
    tekshirardi: omborchi o'z login-parolini kiritib, o'z amalini o'zi
    "tasdiqlab" qo'yardi. Ikki kishilik nazorat umuman ishlamasdi.
    """
    if username == requester.username:
        raise HTTPException(
            status_code=403,
            detail="Amalni o'zingiz tasdiqlay olmaysiz — menejer yoki admin tasdig'i kerak",
        )
    return await verify_approver(db, requester, username, password)

@router.get("/", response_model=List[SupplierOut])
async def get_suppliers(
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role not in ["admin", "manager", "warehouse"]:
        raise HTTPException(status_code=403, detail="Ruxsat berilmagan")
    result = await db.execute(select(Supplier).order_by(Supplier.name))
    return result.scalars().all()

@router.post("/", response_model=SupplierOut)
async def create_supplier(
    supplier: SupplierCreate, 
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role not in ["admin", "manager", "warehouse"]:
        raise HTTPException(status_code=403, detail="Ruxsat berilmagan")
    db_supplier = Supplier(**supplier.model_dump())
    db.add(db_supplier)
    await db.commit()
    await db.refresh(db_supplier)
    
    await log_action(db, current_user.id, "YANGI_FIRMA", f"Firma qo'shildi: {db_supplier.name} (ID: {db_supplier.id})")
    await db.commit() # Commit again to save log
    
    return db_supplier

@router.post("/receipts")
async def add_receipt(
    supplier_id: int = Form(...),
    # Manfiy yoki NaN summa firma balansini qaytarib bo'lmaydigan darajada
    # buzardi (NaN butun "Firmalar" sahifasini 500 ga olib borardi). Bu — pul
    # bilan ishlaydigan yagona joy edi, unga son cheklovlari yetib bormagan.
    total_amount: float = Form(..., gt=0, allow_inf_nan=False),
    note: Optional[str] = Form(None),
    image: Optional[UploadFile] = File(None),
    confirm_username: str = Form(...),
    confirm_password: str = Form(...),
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role not in ["admin", "manager", "warehouse"]:
        raise HTTPException(status_code=403, detail="Ruxsat berilmagan")
    confirming_employee = await verify_confirming_employee(db, current_user, confirm_username, confirm_password)
    # Check if supplier exists
    res = await db.execute(select(Supplier).where(Supplier.id == supplier_id))
    supplier = res.scalars().first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Firma topilmadi")

    image_path = None
    if image:
        # Kengaytmani foydalanuvchi bergan fayl nomidan olardik va uni
        # o'zgartirmasdan saqlardik. /uploads papkasi ilovaning O'Z domenidan
        # beriladi, shuning uchun ".html" yuklab, keyin o'sha havolani ochgan
        # xodimning tokenini o'g'irlash mumkin edi. Endi faqat rasm va PDF.
        ext = os.path.splitext(image.filename or "")[1].lower()
        if ext not in ALLOWED_INVOICE_EXTENSIONS:
            raise HTTPException(
                status_code=400,
                detail=(
                    "Nakladnoy uchun faqat rasm yoki PDF yuklash mumkin "
                    f"({', '.join(sorted(ALLOWED_INVOICE_EXTENSIONS))})"
                ),
            )

        payload = await image.read()
        if len(payload) > MAX_INVOICE_BYTES:
            raise HTTPException(
                status_code=400,
                detail=f"Fayl juda katta (maksimum {MAX_INVOICE_BYTES // (1024 * 1024)} MB)",
            )

        filename = f"{uuid.uuid4()}{ext}"
        save_dir = UPLOAD_DIR / "invoices"
        save_dir.mkdir(parents=True, exist_ok=True)
        (save_dir / filename).write_bytes(payload)

        # Use URL path for DB
        image_path = f"/uploads/invoices/{filename}"

    # Create receipt
    receipt = SupplyReceipt(
        supplier_id=supplier_id,
        total_amount=total_amount,
        invoice_image=image_path,
        note=note
    )
    db.add(receipt)
    
    # Firma balansini ATOMIK oshiramiz (qarzimiz ortadi).
    await db.execute(
        update(Supplier)
        .where(Supplier.id == supplier_id)
        .values(balance=Supplier.balance + total_amount)
        .execution_options(synchronize_session=False)
    )
    
    await log_action(db, current_user.id, "FIRMA_KIRIM", f"Firma: {supplier.name}. Summa: {total_amount} so'm. Izoh: {note or '-'}. Tasdiqladi: @{confirming_employee.username}")
    
    await db.commit()
    await db.refresh(supplier)
    return {"message": "Kirim muvaffaqiyatli saqlandi", "new_balance": supplier.balance}

@router.post("/payments")
async def add_payment(
    supplier_id: int = Form(...),
    amount: float = Form(..., gt=0, allow_inf_nan=False),
    payment_method: str = Form("cash"),
    note: Optional[str] = Form(None),
    confirm_username: str = Form(...),
    confirm_password: str = Form(...),
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role not in ["admin", "manager", "warehouse"]:
        raise HTTPException(status_code=403, detail="Ruxsat berilmagan")
    confirming_employee = await verify_confirming_employee(db, current_user, confirm_username, confirm_password)
    res = await db.execute(select(Supplier).where(Supplier.id == supplier_id))
    supplier = res.scalars().first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Firma topilmadi")

    payment = SupplierPayment(
        supplier_id=supplier_id,
        amount=amount,
        payment_method=payment_method,
        note=note
    )
    db.add(payment)
    
    # Firma balansini ATOMIK kamaytiramiz (qarzimiz kamayadi).
    await db.execute(
        update(Supplier)
        .where(Supplier.id == supplier_id)
        .values(balance=Supplier.balance - amount)
        .execution_options(synchronize_session=False)
    )
    
    await log_action(db, current_user.id, "FIRMA_TOLOV", f"Firma: {supplier.name}. Summa: {amount} so'm. Usul: {payment_method}. Izoh: {note or '-'}. Tasdiqladi: @{confirming_employee.username}")
    
    await db.commit()
    await db.refresh(supplier)
    return {"message": "To'lov muvaffaqiyatli saqlandi", "new_balance": supplier.balance}

@router.get("/{supplier_id}/history")
async def get_supplier_history(
    supplier_id: int, 
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role not in ["admin", "manager", "warehouse"]:
        raise HTTPException(status_code=403, detail="Ruxsat berilmagan")
    # Get receipts
    receipts_res = await db.execute(
        select(SupplyReceipt).where(SupplyReceipt.supplier_id == supplier_id).order_by(desc(SupplyReceipt.date))
    )
    receipts = receipts_res.scalars().all()

    # Get payments
    payments_res = await db.execute(
        select(SupplierPayment).where(SupplierPayment.supplier_id == supplier_id).order_by(desc(SupplierPayment.date))
    )
    payments = payments_res.scalars().all()

    # Merge and sort
    history = []
    for r in receipts:
        history.append({
            "type": "receipt",
            "id": r.id,
            "amount": r.total_amount,
            "date": r.date,
            "image": r.invoice_image,
            "note": r.note
        })
    for p in payments:
        history.append({
            "type": "payment",
            "id": p.id,
            "amount": p.amount,
            "date": p.date,
            "method": p.payment_method,
            "note": p.note
        })
    
    history.sort(key=lambda x: x["date"], reverse=True)
    return history

File: php/app/routes/suppliers.php
<?php
/**
 * /suppliers — firmalar, kirim nakladnoylari, firmaga to'lovlar.
 *
 * PUL MODELI: Supplier.balance MUSBAT bo'lsa do'kon firmaga QARZDOR.
 * Kirim balansni oshiradi, to'lov kamaytiradi.
 *
 * IKKI KISHILIK NAZORAT: har bir pul harakati BOSHQA xodimning (admin yoki
 * menejer) login-paroli bilan tasdiqlanadi. Ilgari bu tekshiruv na rolni,
 * na tasdiqlovchi boshqa odam ekanini ko'rardi — omborchi o'z amalini o'zi
 * "tasdiqlab" qo'yardi va nazorat umuman ishlamasdi.
 */

declare(strict_types=1);

// Nakladnoy fayli uchun ruxsat etilgan kengaytmalar va hajm chegarasi.
const ALLOWED_INVOICE_EXTENSIONS = ['jpg', 'jpeg', 'png', 'webp', 'gif', 'pdf'];
const MAX_INVOICE_BYTES = 10485760; // 10 MB

/**
 * Firma bilan pul harakatini TASDIQLOVCHI xodim.
 * O'zini o'zi tasdiqlashga yo'l qo'ymaydi.
 */
function verify_confirming_employee(array $requester, ?string $username, ?string $password): array
{
    if ($username !== null && $username === $requester['username']) {
        fail(403, "Amalni o'zingiz tasdiqlay olmaysiz — menejer yoki admin tasdig'i kerak");
    }
    return Auth::verifyApprover($requester, $username, $password);
}

// =======================================================================
// GET /suppliers/
// =======================================================================
Router::get('/', function (): never {
    Auth::require(['admin', 'manager', 'warehouse'], 'Ruxsat berilmagan');
    Http::json(array_map(
        fn($r) => Shape::supplier($r),
        Db::all('SELECT * FROM suppliers ORDER BY name')
    ));
});

// =======================================================================
// POST /suppliers/
// =======================================================================
Router::post('/', function (): never {
    $user = Auth::require(['admin', 'manager', 'warehouse'], 'Ruxsat berilmagan');
    $b = Http::body();

    $name = Http::reqStr($b, 'name', 200, 'Nomi');
    $data = [
        'name'       => $name,
        'phone'      => Http::str($b, 'phone', null, 30, 'Telefon'),
        'address'    => Http::str($b, 'address', null, 300, 'Manzil'),
        'balance'    => 0.0,
        'created_at' => Tz::now(),
    ];

    $id = Db::tx(function () use ($data, $user, $name): int {
        $id = Db::insert('suppliers', $data);
        Audit::log((int)$user['id'], 'YANGI_FIRMA', "Firma qo'shildi: $name (ID: $id)");
        return $id;
    });

    Http::json(Shape::supplier(Db::one('SELECT * FROM suppliers WHERE id = ?', [$id])));
});

// =======================================================================
// POST /suppliers/receipts  (multipart: nakladnoy rasmi bilan)
// =======================================================================
Router::post('/receipts', function (): never {
    $user = Auth::require(['admin', 'manager', 'warehouse'], 'Ruxsat berilmagan');
    $f = Http::form();

    $supplierId = Http::reqInt($f, 'supplier_id', 'Firma');
    // Manfiy yoki NaN summa firma balansini qaytarib bo'lmaydigan darajada
    // buzardi (NaN butun "Firmalar" sahifasini 500 ga olib borardi).
    $totalAmount = Http::reqNum($f, 'total_amount', 'gt', 0, null, 'Summa');
    $note = Http::str($f, 'note', null, 500, 'Izoh');

    $confirming = verify_confirming_employee(
        $user,
        Http::reqStr($f, 'confirm_username', 150, 'Tasdiqlovchi login'),
        Http::reqStr($f, 'confirm_password', 200, 'Tasdiqlovchi parol')
    );

    $supplier = Db::one('SELECT * FROM suppliers WHERE id = ?', [$supplierId]);
    if ($supplier === null) {
        fail(404, 'Firma topilmadi');
    }

    $imagePath = save_invoice_image();

    Db::tx(function () use ($supplierId, $totalAmount, $imagePath, $note, $user, $supplier, $confirming): void {
        Db::insert('supply_receipts', [
            'supplier_id'   => $supplierId,
            'total_amount'  => $totalAmount,
            'invoice_image' => $imagePath,
            'date'          => Tz::now(),
            'note'          => $note,
        ]);

        // Firma balansini ATOMIK oshiramiz (qarzimiz ortadi).
        Db::run('UPDATE suppliers SET balance = balance + ? WHERE id = ?', [$totalAmount, $supplierId]);

        Audit::log((int)$user['id'], 'FIRMA_KIRIM',
            "Firma: {$supplier['name']}. Summa: $totalAmount so'm. Izoh: "
            . ($note ?? '-') . '. Tasdiqladi: @' . $confirming['username']);
    });

    $fresh = Db::one('SELECT balance FROM suppliers WHERE id = ?', [$supplierId]);
    Http::json([
        'message'     => 'Kirim muvaffaqiyatli saqlandi',
        'new_balance' => Db::f($fresh['balance']),
    ]);
});

/**
 * Nakladnoy faylini saqlaydi va URL yo'lini qaytaradi.
 *
 * Kengaytmani ilgari foydalanuvchi bergan fayl nomidan olardik va uni
 * o'zgartirmasdan saqlardik. /uploads papkasi ilovaning O'Z domenidan
 * beriladi, shuning uchun ".html" yuklab, keyin o'sha havolani ochgan
 * xodimning tokenini o'g'irlash mumkin edi. Endi faqat rasm va PDF, nomi
 * esa tasodifiy — foydalanuvchi bergan nom umuman ishlatilmaydi.
 */
function save_invoice_image(): ?string
{
    if (!isset($_FILES['image']) || ($_FILES['image']['error'] ?? UPLOAD_ERR_NO_FILE) === UPLOAD_ERR_NO_FILE) {
        return null;
    }

    $file = $_FILES['image'];
    if ($file['error'] !== UPLOAD_ERR_OK) {
        if ($file['error'] === UPLOAD_ERR_INI_SIZE || $file['error'] === UPLOAD_ERR_FORM_SIZE) {
            fail(400, 'Fayl juda katta (maksimum ' . (MAX_INVOICE_BYTES / 1048576) . ' MB)');
        }
        fail(400, 'Faylni yuklashda xatolik.');
    }

    if ($file['size'] > MAX_INVOICE_BYTES) {
        fail(400, 'Fayl juda katta (maksimum ' . (MAX_INVOICE_BYTES / 1048576) . ' MB)');
    }

    $ext = strtolower(pathinfo((string)($file['name'] ?? ''), PATHINFO_EXTENSION));
    if (!in_array($ext, ALLOWED_INVOICE_EXTENSIONS, true)) {
        fail(400, 'Nakladnoy uchun faqat rasm yoki PDF yuklash mumkin ('
            . implode(', ', ALLOWED_INVOICE_EXTENSIONS) . ')');
    }

    // Mazmunini ham tekshiramiz: kengaytmani almashtirib qo'yish oson.
    if (function_exists('finfo_open')) {
        $finfo = finfo_open(FILEINFO_MIME_TYPE);
        if ($finfo !== false) {
            $mime = (string)finfo_file($finfo, $file['tmp_name']);
            finfo_close($finfo);
            $okMimes = ['image/jpeg', 'image/png', 'image/webp', 'image/gif', 'application/pdf'];
            if (!in_array($mime, $okMimes, true)) {
                fail(400, "Fayl mazmuni rasm yoki PDF emas.");
            }
        }
    }

    $saveDir = UPLOAD_DIR . '/invoices';
    if (!is_dir($saveDir) && !@mkdir($saveDir, 0755, true) && !is_dir($saveDir)) {
        fail(500, "Fayl saqlash papkasini yaratib bo'lmadi.");
    }

    try {
        $filename = bin2hex(random_bytes(16)) . '.' . $ext;
    } catch (Throwable) {
        $filename = uniqid('inv', true) . '.' . $ext;
    }

    if (!move_uploaded_file($file['tmp_name'], "$saveDir/$filename")) {
        fail(500, "Faylni saqlab bo'lmadi.");
    }
    @chmod("$saveDir/$filename", 0644);

    return "/uploads/invoices/$filename";
}

// =======================================================================
// POST /suppliers/payments
// =======================================================================
Router::post('/payments', function (): never {
    $user = Auth::require(['admin', 'manager', 'warehouse'], 'Ruxsat berilmagan');
    $f = Http::form();

    $supplierId = Http::reqInt($f, 'supplier_id', 'Firma');
    $amount = Http::reqNum($f, 'amount', 'gt', 0, null, 'Summa');
    $method = Http::str($f, 'payment_method', 'cash', 40, "To'lov usuli") ?? 'cash';
    $note = Http::str($f, 'note', null, 500, 'Izoh');

    $confirming = verify_confirming_employee(
        $user,
        Http::reqStr($f, 'confirm_username', 150, 'Tasdiqlovchi login'),
        Http::reqStr($f, 'confirm_password', 200, 'Tasdiqlovchi parol')
    );

    $supplier = Db::one('SELECT * FROM suppliers WHERE id = ?', [$supplierId]);
    if ($supplier === null) {
        fail(404, 'Firma topilmadi');
    }

    Db::tx(function () use ($supplierId, $amount, $method, $note, $user, $supplier, $confirming): void {
        Db::insert('supplier_payments', [
            'supplier_id'    => $supplierId,
            'amount'         => $amount,
            'payment_method' => $method,
            'date'           => Tz::now(),
            'note'           => $note,
        ]);

        // Firma balansini ATOMIK kamaytiramiz (qarzimiz kamayadi).
        Db::run('UPDATE suppliers SET balance = balance - ? WHERE id = ?', [$amount, $supplierId]);

        Audit::log((int)$user['id'], 'FIRMA_TOLOV',
            "Firma: {$supplier['name']}. Summa: $amount so'm. Usul: $method. Izoh: "
            . ($note ?? '-') . '. Tasdiqladi: @' . $confirming['username']);
    });

    $fresh = Db::one('SELECT balance FROM suppliers WHERE id = ?', [$supplierId]);
    Http::json([
        'message'     => "To'lov muvaffaqiyatli saqlandi",
        'new_balance' => Db::f($fresh['balance']),
    ]);
});

// =======================================================================
// GET /suppliers/{supplier_id}/history
// =======================================================================
Router::get('/{supplier_id}/history', function (array $p): never {
    Auth::require(['admin', 'manager', 'warehouse'], 'Ruxsat berilmagan');
    $supplierId = Router::id($p, 'supplier_id');

    $receipts = Db::all(
        'SELECT * FROM supply_receipts WHERE supplier_id = ? ORDER BY date DESC, id DESC LIMIT 1000',
        [$supplierId]
    );
    $payments = Db::all(
        'SELECT * FROM supplier_payments WHERE supplier_id = ? ORDER BY date DESC, id DESC LIMIT 1000',
        [$supplierId]
    );

    $history = [];
    foreach ($receipts as $r) {
        $history[] = [
            'type'   => 'receipt',
            'id'     => Db::i($r['id']),
            'amount' => Db::f($r['total_amount']),
            'date'   => Tz::iso($r['date']),
            'image'  => $r['invoice_image'] ?? null,
            'note'   => $r['note'] ?? null,
            '_sort'  => (string)$r['date'],
        ];
    }
    foreach ($payments as $r) {
        $history[] = [
            'type'   => 'payment',
            'id'     => Db::i($r['id']),
            'amount' => Db::f($r['amount']),
            'date'   => Tz::iso($r['date']),
            'method' => $r['payment_method'],
            'note'   => $r['note'] ?? null,
            '_sort'  => (string)$r['date'],
        ];
    }

    usort($history, fn($a, $b) => strcmp($b['_sort'], $a['_sort']));
    foreach ($history as &$h) {
        unset($h['_sort']);
    }

    Http::json($history);
});
